### DISCUSSION TITLE
Tracking warnings and deprecations

### DISCUSSION THREAD

**Discussion by ananthsub** [Lightning-AI/pytorch-lightning#8878] [General]
What's the recommendation for how the project should track these?

In some parts of the codebase, we use `rank_zero_warn`/`rank_zero_deprecation`
In other parts of the codebase we use the warning cache 
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/warnings.py

Why do we have both? When should we use one over another?

And for deprecated functions or classes, should we use a decorator for them? https://pypi.org/project/Deprecated/

cc @Borda @awaelchli @carmocca @tchaton @edward-io

**Comment by awaelchli** [comment]
Not sure myself, I will let @carmocca and @Borda deliver the official answer :) 
My understanding:

rank_zero_warn: simple to use, when we know for sure it will print only once
warning_cache: when we are in a loop and prevent repeated warnings. needs to be assigned as an attribute or global variable within the module however.

I believe we used to have this extension https://pypi.org/project/pyDeprecate/

**Reply by carmocca** [reply]
`rank_zero_warn`. Calls `warnings.warn` with the rank zero wrapper. `warnings.warn` is supposed to print the same warning only once in the process.

`WarningCache`. Calls `rank_zero_warn` internally. I believe it was added because the "once in the process" requirement is unfulfilled in some distributed setting (i'm not sure if it's dp or what, but it was added with https://github.com/PyTorchLightning/pytorch-lightning/pull/3618). Maybe worth testing when it's necessary exactly and writing a test for that.

 > I believe we used to have this extension https://pypi.org/project/pyDeprecate/

We still install it:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/requirements.txt#L11

Although we pretty much don't use it now that torchmetrics has been removed from the project.

```python
$ grep -iIrn "from deprecate" pytorch_lightning
pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:18:from deprecate import void
pytorch_lightning/loops/epoch/prediction_epoch_loop.py:4:from deprecate import void
pytorch_lightning/loops/batch/training_batch_loop.py:23:from deprecate import void
pytorch_lightning/loops/base.py:18:from deprecate import void
pytorch_lightning/loops/dataloader/evaluation_loop.py:17:from deprecate.utils import void
pytorch_lightning/loops/dataloader/prediction_loop.py:3:from deprecate.utils import void
```
